pkg/avx/collector.go

```diff
@@ -77,111 +102,123 @@ var (
+		return -1, errors.New("unable to set up perf events")
 	}
 
-	return nil
-}
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_SET_BPF, uintptr(prog.FD())); errno != 0 {
+		return -1, errors.New("unable to attach bpf program to perf events")
+	}
 
-type collector struct {
-	root                     string
-	elfFilepath              string
-	bpfModule                *bpf.Module
-	avxContextSwitchCounters *bpf.Map
-	allContextSwitchCounters *bpf.Map
-	lastUpdateNs             *bpf.Map
-	lastCPUCounters          *bpf.Map
+	return pfd, nil
 }
 
-// NewCollector creates new Prometheus collector for AVX metrics
-func NewCollector() (prometheus.Collector, error) {
+func getKernelVersion() uint32 {
 
-	elfFilepath := path.Join(bpfInstallpath, bpfBinaryName)
+	var uts unix.Utsname
 
-	if err := checkElfKernelVersion(elfFilepath); err != nil {
-		return nil, err
+	err := unix.Uname(&uts)
+	if err != nil {
+		return 0
 	}
 
-	bpfModule := bpf.NewModule(elfFilepath)
+	str := string(bytes.SplitN(uts.Release[:], []byte{0}, 2)[0])
+
+	ver := strings.SplitN(str, ".", 3)
 
-	sectionParams := make(map[string]bpf.SectionParams)
-	if err := bpfModule.Load(sectionParams); err != nil {
-		return nil, errors.Wrap(err, "unable to load eBPF ELF file")
+	major, err := strconv.ParseUint(ver[0], 10, 8)
+	if err != nil {
+		return 0
+	}
+	minor, err := strconv.ParseUint(ver[1], 10, 8)
+	if err != nil {
+		return uint32(major << 16)
 	}
 
-	allSwitchCounters := bpfModule.Map("all_context_switch_count")
-	if allSwitchCounters == nil {
-		return nil, errors.New("map all_context_switch_count not found")
+	// ignore patch version
+	return uint32(major<<16 + minor<<8)
+}
+
+// NewCollector creates new Prometheus collector for AVX metrics
+func NewCollector() (prometheus.Collector, error) {
+
+	// Set rlimit to be able to lock map values in memory
+	memlockLimit := &unix.Rlimit{
+		Cur: mapMemLockLimit,
+		Max: mapMemLockLimit,
+	}
+	err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, memlockLimit)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to set rlimit")
 	}
 
-	avxSwitchCounters := bpfModule.Map("avx_context_switch_count")
-	if avxSwitchCounters == nil {
-		return nil, errors.New("map avx_context_switch_count not found")
+	spec, err := bpf.LoadCollectionSpec(filepath.Join(bpfInstallpath, bpfBinaryName))
+	if err != nil {
+		log.Info("Unable to load user eBPF (%v). Using default CollectionSpec from ELF program bytes", err)
+		spec, err = bpf.LoadCollectionSpecFromReader(bytes.NewReader(program[:]))
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to load default CollectionSpec from ELF program bytes")
+		}
 	}
 
-	lastUpdateNs := bpfModule.Map("last_update_ns")
-	if lastUpdateNs == nil {
-		return nil, errors.New("map last_update_ns not found")
+	ver := getKernelVersion()
+
+	for _, prog := range spec.Programs {
+		if ver < prog.KernelVersion {
+			return nil, errors.Wrap(err, "host kernel too old to run the program")
```

Comment:
nit: I'd suggest also printing that minimal (prog.KernelVersion) acceptable version to let the user know.

Actually, and this is purely by looking at the code and without really understanding whether it can happen in practice that different spec.Programs have different kernel version requirements, I'd change that loop to not break out on the firsts version check failure but instead keep going and report all. You know, just to avoid hoards of angry sysadmins putting out a contract on our head once, after the 10th consecutive kernel upgrade, cri-resmgr still refuses to start up demanding one more newer kernel. I'd probably just turn that loop into a dedicated kernel version check function that return nil on success and an error on failure which would include the *highest* required kernel version among spec.Programs...

Reply:
there's only one version specified in the elf so with the avx512 program(s), we can exit immediately. I made it in a loop to avoid the need to hard code map keys...

Reply:
Hmm... I meant potentially multiple loaded ELF binaries in the AVX collector in the future, which `ebpf.CollectionSpec` seems to allow. But if we never going to have/need multiple programs then breaking out is fine. But could you add that required `prog.KernelVersion` to the returned error ?

Reply:
> which `ebpf.CollectionSpec` seems to allow

AFAICS the elf reader gives one `CollectionSpec` per elf file. An elf can specify multiple programs. 

I've added the version info to the error for both minimum version and host kernel version .

Reply:
This LGTM enough to merge and then address any remaining issues with further patches/PRs.